src/cargo/ops/cargo_add/mod.rs

```diff
@@ -0,0 +1,585 @@
+
+    Ok(dependency)
+}
+
+/// Provide the existing dependency for the target table
+///
+/// If it doesn't exist but exists in another table, let's use that as most likely users
+/// want to use the same version across all tables unless they are renaming.
+fn get_existing_dependency(
+    manifest: &LocalManifest,
+    dep_key: &str,
+    section: DepTable<'_>,
+) -> Option<Dependency> {
+    #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+    enum Key {
+        Dev,
+        Build,
+        Target,
+        Runtime,
+        Existing,
+    }
+
+    let target_section = section.to_table();
+    let mut possible: Vec<_> = manifest
+        .get_dependency_versions(dep_key)
+        .filter_map(|(path, dep)| dep.ok().map(|dep| (path, dep)))
+        .map(|(path, dep)| {
+            let key = if path == target_section {
+                Key::Existing
+            } else {
+                match path[0].as_str() {
+                    "dependencies" => Key::Runtime,
+                    "target" => Key::Target,
+                    "build-dependencies" => Key::Build,
+                    "dev-dependencies" => Key::Dev,
+                    other => unreachable!("Unknown dependency section: {}", other),
+                }
+            };
+            (key, dep)
+        })
+        .collect();
+    possible.sort_by_key(|(key, _)| *key);
+    let (key, mut dep) = possible.pop()?;
+
+    if key != Key::Existing {
+        // When the dep comes from a different section, we only care about the source and not any
+        // of the other fields, like `features`
+        let unrelated = dep;
+        dep = Dependency::new(&unrelated.name);
+        dep.source = unrelated.source.clone();
+        dep.registry = unrelated.registry.clone();
+
+        // dev-dependencies do not need the version populated when path is set though we
+        // should preserve it if the user chose to populate it.
+        let version_required = unrelated.source().and_then(|s| s.as_registry()).is_some();
+        let version_optional_in_section = section == DepTable::Development;
+        if !version_required && version_optional_in_section {
+            dep = dep.clear_version();
+        }
+    }
+
+    Some(dep)
+}
+
+fn get_latest_dependency(
+    dependency: &Dependency,
+    _flag_allow_prerelease: bool,
+    config: &Config,
+    registry: &mut crate::core::registry::PackageRegistry<'_>,
+) -> CargoResult<Dependency> {
+    let query = dependency.query(config)?;
+    let possibilities = loop {
+        match registry.query_vec(&query, true) {
+            std::task::Poll::Ready(res) => {
+                break res?;
+            }
+            std::task::Poll::Pending => registry.block_until_ready()?,
+        }
+    };
+    let latest = possibilities
+        .iter()
+        .max_by_key(|s| {
+            let stable = s.version().pre.is_empty();
+            (stable, s.version())
+        })
```

Comment:
If I'm reading this correctly:
- `semver` says `1.0.0 < 2.0.0-alpha.1`
- `cargo-add` says `2.0.0-alpha.1 < 1.0.0`

The goal with this code was to only select a pre-release if that is the only option available.

Reply:
I am reading the semver crate again. It does state that pre-release versions will be ignored if not specified in `VersionReq` in the doc and its code. Also, each implementation of `Source::query` calls `Dependency::matches_id` and eventually calls `VersionReq::matches`. That being said,

- If prerelease is specified in `VersionReq`, that means the user is fine with picking prerelease version, and semver will pick them if possible.
- If not, semver just ignores them.

Maybe we needn't to check prerelease here and depend on semver for consistency?

- https://github.com/dtolnay/semver/blob/8d8bdb2adaf4572b3c4e8502e7498838b2cf7ccb/src/eval.rs#L3-L24
- https://docs.rs/semver/1.0.6/semver/struct.VersionReq.html#associatedconstant.STAR
- https://github.com/rust-lang/cargo/blob/b816d82ee94acd4d6be588c0f0cd299701c55457/src/cargo/core/source/mod.rs#L32

Reply:
Maybe we can add a package with version `100000.0.0+my-package` to our tests to prevent regression. The test case [`install::pick_max_version`](https://github.com/rust-lang/cargo/blob/25fcb135d02ea897ce894b67ae021f48107d522b/tests/testsuite/install.rs#L271-L294) already did that.

Reply:
The important thing to remember is this function's intention is for us to find a version for creating a version-req.  We are in the case where `dep.source().is_none()`.  For example, if you do a `cargo add clap` on an empty crate, we should select 3.1.6 but no pre-release version.  The user can explicitly do so with `cargo add clap@3.0.0-rc.0`.  However, if the crate only exists as pre-release versions, we should allow the user to add that.

Reply:
> Maybe we can add a package with version 100000.0.0+my-package to our tests to prevent regression. The test case [install::pick_max_version](https://github.com/rust-lang/cargo/blob/25fcb135d02ea897ce894b67ae021f48107d522b/tests/testsuite/install.rs#L271-L294) already did that.

`add_basic` (and many tests using the common package names) currently select from
- 0.1.1
- 0.2.3
- 0.4.1
- 20.0.0
- 999999.0.0 (no idea if I have the number of 9s right)
- 999999.0.0-alpha.1 

Unsure what 100000.0.0+my-package would add that isn't already handled.

I am added a test case that is pre-release only

Reply:
Ah! I think I got the difference. So basically cargo-add tries harder than cargo-install on finding a usable version. cargo-install avoids prerelease version by [passing a `STAR` VersionReq](https://github.com/rust-lang/cargo/blob/b816d82ee94acd4d6be588c0f0cd299701c55457/src/cargo/ops/cargo_install.rs#L81-L90), whereas cargo-add [gets a `Dependency` with `OptVersionReq::Any` which accepts any versions](https://github.com/rust-lang/cargo/blob/c5f33768a52cc7cffe620ca397ec0fd3b3d2b879/src/cargo/ops/cargo_add/dependency.rs#L213) if unspecified.

Anyway, thanks for your explanation. I am fine with this tiny difference and it indeed makes cargo-add more convenient. 😄